File: 2022/P21/problem21a.py
data = []
with open("problem21.txt", 'r') as f:
    data = f.read().strip('\n').split('\n')

data = [x.split(': ') for x in data]

ops = {'+': lambda x, y: x + y, '-': lambda x, y: x - y, '*': lambda x, y: x * y, '/': lambda x, y: x // y}

new_data = {}
for d in data:
    d1 = d[0]
    d2 = d[1]
    for c in ops:
        if c in d2:
            p1, p2 = d2.split(' ' + c + ' ')
            new_data[d1] = (c, p1, p2)
            break
    else:
        new_data[d1] = (int(d2),)

def calculate_monkey_tree(m, d):
    val = d[m]
    if val[0] in ops:
        f = ops[val[0]]
        t1 = calculate_monkey_tree(val[1], d)
        t2 = calculate_monkey_tree(val[2], d)
        return f(t1, t2)
    else:
        return val[0]
    
print(calculate_monkey_tree('root', new_data))

File: 2023/P21/problem21a.rs
use std::fs;
use std::collections::VecDeque;

fn main() {
    let data = fs::read_to_string("problem21.txt")
        .expect("Failed to read file");
    let data = data
        .trim()
        .split('\n')
        .collect::<Vec<_>>();
    let data = data
        .iter()
        .map(|line| {
            let mut char_vec = Vec::<char>::new();
            for c in line.chars() {
                char_vec.push(c);
            }
            char_vec
        })
        .collect::<Vec<_>>();
    let mut visited = Vec::<Vec<Option<u32>>>::new();
    for line in data.iter() {
        let mut visited_vec = Vec::<Option<u32>>::new();
        for _ in line {
            visited_vec.push(None);
        }
        visited.push(visited_vec);
    }
    let mut start_point = None;
    for i in 0..data.len() {
        for j in 0..data[0].len() {
            if data[i][j] == 'S' {
                start_point = Some((i, j));
            }
        }
    }
    let start_point = start_point.expect("Error: Couldn't find S in grid");
    let mut q = VecDeque::<(u32, (usize, usize))>::new();
    q.push_back((0, start_point));
    while let Some((
        dist,
        curr_pos
    )) = q.pop_front() {
        if visited[curr_pos.0][curr_pos.1] != None {
            continue;
        }
        visited[curr_pos.0][curr_pos.1] = Some(dist);
        for dir in [(0, 1), (1, 0), (0, -1), (-1, 0)] {
            let next_pos = ((curr_pos.0 as i32) + dir.0, (curr_pos.1 as i32) + dir.1);
            match next_pos {
                (x, y) if (x >= 0
                    && x < data.len() as i32
                    && y >= 0
                    && y < data.len() as i32
                    && data[x as usize][y as usize] != '#') => {
                    q.push_back((dist + 1, (x as usize, y as usize)));
                },
                _ => { }
            }
        }
    }
    let mut sum = 0;
    for line in visited {
        for v in line {
            match v {
                Some(x) if x % 2 == 0 && x <= 64 => {
                    sum += 1;
                },
                _ => { }
            }
        }
    }
    println!("{sum}");
}
